src/View/Helper/PaginatorHelper.php

```diff
@@ -313,9 +313,12 @@ protected function _toggledLink($text, $enabled, $options, $templates): string
         }
         $paging = $this->params($options['model']);
 
-        $url = array_merge(
+        if (!isset($options['url']['?'])) {
+            $options['url']['?'] = [];
+        }
+        $url = Hash::merge(
             $options['url'],
-            ['page' => $paging['page'] + $options['step']]
+            ['?' => ['page' => $paging['page'] + $options['step']]]
```

Comment:
This will have the drawback of making application routes not able to place page and sort parameters in the path.

Reply:
True, but now we can't simply merge as before either and it would cause route mismatch by default when you haven't setup any routes with the paging params as placeholders.

Solution would be to allow configuring which paging params should be treated as routes placeholders.